src/base/USong.pas

```diff
@@ -796,281 +799,328 @@ function TSong.ReadTXTHeader(SongFile: TTextFileStream; ReadCustomTags: Boolean)
+    end;
+    
+    // Song Ending
+    if (TagMap.TryGetData('END', Value)) then
+    begin
+      TagMap.Remove('END');
+      TryStrtoInt(Value, self.Finish)
+    end;
+    
+    // Resolution
+    if (TagMap.TryGetData('RESOLUTION', Value)) then
+    begin
+      TagMap.Remove('RESOLUTION');
+      TryStrtoInt(Value, self.Resolution);
+      if (self.Resolution < 1) then
       begin
-        AddCustomTag(Identifier, Value);
+        Log.LogError('Ignoring invalid resolution in song: ' + FullFileName);
+        self.Resolution := DEFAULT_RESOLUTION;
       end;
-    end; // End check for non-empty Value
-
-    // read next line
-    if not SongFile.ReadLine(Line) then
+    end;
+    
+    // Notes Gap
+    if (TagMap.TryGetData('NOTESGAP', Value)) then
     begin
-      Result := false;
-      Log.LogError('File incomplete or not Ultrastar txt (A): ' + FullFileName);
-      Break;
+      TagMap.Remove('NOTESGAP');
+      TryStrtoInt(Value, self.NotesGAP)
     end;
-  end; // while
+    
+    // Relative Notes
+    if (TagMap.TryGetData('RELATIVE', Value)) then
+    begin
+      TagMap.Remove('RELATIVE');
+      if (UpperCase(Value) = 'YES') then
+        self.Relative := true;
+    end;
+    
+    // File encoding
+    if (TagMap.TryGetData('ENCODING', Value)) then
+    begin
+      TagMap.Remove('ENCODING');
+      self.Encoding := ParseEncoding(Value, Ini.DefaultEncoding);
+    end;
+    
+    // PreviewStart
+    if (TagMap.TryGetData('PREVIEWSTART', Value)) then
+    begin
+      TagMap.Remove('PREVIEWSTART');
+      self.PreviewStart := StrToFloatI18n( Value );
+      if (self.PreviewStart>0) then
+      begin
+        MedleyFlags := MedleyFlags or 1;
+        HasPreview := true;
+      end;
+    end;
+    
+    // MedleyStartBeat
+    if TagMap.TryGetData('MEDLEYSTARTBEAT', Value) and not self.Relative then
+    begin
+      TagMap.Remove('MEDLEYSTARTBEAT');
+      if TryStrtoInt(Value, self.Medley.StartBeat) then
+        MedleyFlags := MedleyFlags or 2;
+    end;
+    
+    // MedleyEndBeat
+    if TagMap.TryGetData('MEDLEYENDBEAT', Value) and not self.Relative then
+    begin
+      TagMap.Remove('MEDLEYENDBEAT');
+      if TryStrtoInt(Value, self.Medley.EndBeat) then
+        MedleyFlags := MedleyFlags or 4;
+    end;
+    
+    // Medley
+    if (TagMap.TryGetData('CALCMEDLEY', Value)) then
+    begin
+      TagMap.Remove('CALCMEDLEY');
+      if Uppercase(Value) = 'OFF' then
+        self.CalcMedley := false;
+    end;
+    
+    // Duet Singer Name P1
+    if (TagMap.TryGetData('DUETSINGERP1', Value)) then
+    begin
+      TagMap.Remove('DUETSINGERP1');
+      DecodeStringUTF8(Value, DuetNames[0], Encoding);
+    end;
+    
+    // Duet Singer Name P2
+    if (TagMap.TryGetData('DUETSINGERP2', Value)) then
+    begin
+      TagMap.Remove('DUETSINGERP2');
+      DecodeStringUTF8(Value, DuetNames[1], Encoding);
+    end;
+    
+    // Duet Singer Name P1
+    if (TagMap.TryGetData('P1', Value)) then
```

Comment:
This is a slight change in behaviour:
- old code: it would use P1 or DUETSINGERP1, whatever was defined _last_ in the headers
- new code: regardless of its place, P1 always overrides DUETSINGERP1

DUETSINGERP1 is deprecated anyway, this is more a comment that I agree with this new behaviour :+1:

Reply:
Yes. I chose this new behavior as it is according to the [format specification](https://github.com/UltraStar-Deluxe/format/blob/main/spec.md#325-the-duetsingerp1-thru-duetsingerp9-headers), where PX takes precedence over DUETSINGERX.